### PR TITLE
partial_sort_copy bugfix

### LINKED ISSUE #1086: <algorithm>: std::partial_sort_copy performs an unconstrained operation
In:

https://github.com/microsoft/STL/blob/c10ae01b4d9508eed9d5f059a120ee7223b6ac12/stl/inc/algorithm#L6946-L6972

on line 6967:

https://github.com/microsoft/STL/blob/c10ae01b4d9508eed9d5f059a120ee7223b6ac12/stl/inc/algorithm#L6967

we construct an object of `_InIt`'s value type from an element reference. The type requirements of the algorithm from [[partial.sort.copy]](http://eel.is/c++draft/partial.sort.copy) are:

> 1. Let N be `min(last - first, result_last - result_first)`. Let `comp` be `less{}`, and `proj1` and `proj2` be `identity{}` for the overloads with no parameters by those names.
> 2. _Mandates_: For the overloads in namespace `std`, the expression `*first` is writable ([[iterator.requirements.general]](http://eel.is/c++draft/iterator.requirements.general)) to `result_first`.
> 3. _Preconditions_: For the overloads in namespace `std`, `RandomAccessIterator` meets the _`Cpp17ValueSwappable`_ requirements ([[swappable.requirements]](http://eel.is/c++draft/swappable.requirements)), the type of `*result_­first` meets the _`Cpp17MoveConstructible`_ ([Table 28](http://eel.is/c++draft/utility.arg.requirements#tab:cpp17.moveconstructible)) and _`Cpp17MoveAssignable`_ ([Table 30](http://eel.is/c++draft/utility.arg.requirements#tab:cpp17.moveassignable)) requirements.

There are requirements that allow the algorithm to construct an object of the random access iterator's value type, but not the input iterator's value type.

To my understanding, this program should compile and run successfully:
```c++
#include <algorithm>
#include <cassert>
#include <compare>

struct wrapper {
    wrapper() = default;
    constexpr explicit wrapper(int i) : x{i} {}

    auto operator<=>(const wrapper&) const = default;

    int x;
};

struct source : wrapper {
    using wrapper::wrapper;

    source(const source&) = delete;
    source& operator=(const source&) = delete;
};

struct target : wrapper {
    using wrapper::wrapper;

    target(target&&) = default;
    target& operator=(target&&) = default;

    target& operator=(const source& w) {
        x = w.x;
        return *this;
    }
};

int main() {
    source src[] = {source{5}, source{4}, source{3}, source{2}, source{1}, source{0}};
    target dst[2];

    using std::begin, std::end;
    std::partial_sort_copy(begin(src), end(src), begin(dst), end(dst));
    assert(dst[0].x == 0);
    assert(dst[1].x == 1);
}
```
but it does not (https://godbolt.org/z/1xq7ah). Interestingly, libstdc++ has the same bug, but libc++ does not. This repro triggers a different `partial_sort_copy` bug in libc++: it uses a default comparator other than `std::less{}` which fails for this test case.

### PR DESCRIPTION
**Pull request by CaseyCarter**
Instead of assuming that the `_Val` argument to the heap helper functions is an rvalue of some type, perfect-forward it all the way into `_Push_heap_by_index` where it's finally written into the hole in the heap. This allows `partial_sort_copy` to pass in a reference to an element in its input range which is then correctly written into the output range in `_Push_heap_by_index`.

I update the `std::ranges` heap helpers similarly, and make their constraints explicit. The helpers with a `_Val` argument now accept a distinct projection to be applied to `_Val`, which will enable the implementation of `std::ranges::partial_sort_copy`.

Drive-by: Replace `_unchecked` in the names of `ranges::_Make_heap_fn::_Make_heap_unchecked` and `ranges::_Sort_heap_fn::_Sort_heap_unchecked` with `_common` to be consistent with the names of other internal helpers that require the sentinel and iterator to have the same type. Promote both functions from class scope to namespace scope as needed to support implementation of `partial_sort` and `partial_sort_copy`.

Fixes #1086.

### CODE REVIEW

File: stl/inc/algorithm [hunk 1]
```diff
@@ -5585,6 +5591,22 @@ _CONSTEXPR20 void make_heap(_RanIt _First, _RanIt _Last) { // make [_First, _Las
 #ifdef __cpp_lib_concepts
 namespace ranges {
     // VARIABLE ranges::make_heap
+    // clang-format off
+    template <random_access_iterator _It, class _Pr, class _Pj>
+        requires sortable<_It, _Pr, _Pj>
+    static constexpr void _Make_heap_common(_It _First, _It _Last, _Pr _Pred, _Pj _Proj) {
+        // make nontrivial [_First, _Last) into a heap with respect to _Pred and _Proj
+        // clang-format on
```
Comment (miscco) [inline stl/inc/algorithm:5599]: That line here is strange, usually we disable whole functions not parts of it

File: tests/std/tests/GH_001086_partial_sort_copy/test.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    source src[4];
+    constexpr auto src_size = static_cast<int>(std::size(src));
+    target dst[2];
+    constexpr auto dst_size = static_cast<int>(std::size(dst));
+
+    for (int i = 0; i < src_size; ++i) {
+        src[i].x = src_size - 1 - i;
+    }
+
+    partial_sort_copy(begin(src), end(src), begin(dst), end(dst));
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_001086_partial_sort_copy/test.cpp:58]: Need to include something more than `<algorithm>` for `begin()/end()`, WG21-N4861 [iterator.range]/1.

File: tests/std/tests/GH_001086_partial_sort_copy/test.cpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        x = w.x;
+        return *this;
+    }
+};
+
+int main() {
+    source src[4];
+    constexpr auto src_size = static_cast<int>(std::size(src));
+    target dst[2];
+    constexpr auto dst_size = static_cast<int>(std::size(dst));
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_001086_partial_sort_copy/test.cpp:52]: `using namespace std;` is above. But actually, it seems unnecessarily complicated to obtain `src_size` and `dst_size` like this. Would it be simpler to say `constexpr int src_size = 4;` followed by `source src[src_size];`?

File: stl/inc/algorithm (hunk 1 above)
Comment (CaseyCarter) [reply stl/inc/algorithm:5599]: Yes, I stumbled across this almost by accident but the intent is that it becomes the new convention. I would swear that I had tried this placement in the past and had poor results - maybe it was with a trailing-requires clause? Anyway, yes, as STL says I'm happy about anything that narrows the scope of code we must format manually.

File: stl/inc/algorithm [hunk 4]
```diff
@@ -5383,7 +5384,7 @@ namespace ranges {
 
             --_Last;
             iter_value_t<_It> _Val = _RANGES iter_move(_Last);
-            _RANGES _Push_heap_by_index(_STD move(_First), _Count - 1, 0, _STD move(_Val), _Pred, _Proj);
+            _RANGES _Push_heap_by_index(_STD move(_First), _Count - 1, 0, _STD move(_Val), _Pred, _Proj, _Proj);
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/algorithm:5387]: I did give this some thought, and I don't believe it's an issue given that only serial algorithms are potentially aliasing function objects so there's no danger of simultaneously invoking the same function object via different indirections. Do you think this merits a call out so future us will know to do something special if we're parallelizing here?

File: stl/inc/algorithm [hunk 5]
```diff
@@ -5818,6 +5824,17 @@ _CONSTEXPR20 void sort_heap(_RanIt _First, _RanIt _Last) { // order heap by repe
 #ifdef __cpp_lib_concepts
 namespace ranges {
     // VARIABLE ranges::sort_heap
+    template <class _It, class _Pr, class _Pj>
+    static constexpr void _Sort_heap_common(const _It _First, _It _Last, _Pr _Pred, _Pj _Proj) {
+        // order heap by repeatedly popping
+        _STL_INTERNAL_STATIC_ASSERT(random_access_iterator<_It>);
+        _STL_INTERNAL_STATIC_ASSERT(sortable<_It, _Pr, _Pj>);
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/algorithm:5831]: Yes, it should, don't know how I missed this one.

File: stl/inc/algorithm (hunk 4 above)
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/algorithm:5887]: I think a short comment is a good idea, yeah.

File: stl/inc/algorithm [hunk 6]
```diff
@@ -6085,6 +6091,22 @@ _CONSTEXPR20 void make_heap(_RanIt _First, _RanIt _Last) { // make [_First, _Las
 #ifdef __cpp_lib_concepts
 namespace ranges {
     // VARIABLE ranges::make_heap
+    // clang-format off
+    template <random_access_iterator _It, class _Pr, class _Pj>
+        requires sortable<_It, _Pr, _Pj>
+    static constexpr void _Make_heap_common(_It _First, _It _Last, _Pr _Pred, _Pj _Proj) {
```
Comment (CaseyCarter) [inline stl/inc/algorithm:6097]: `static` copy-pasta here - please audit.

File: stl/inc/algorithm [hunk 7]
```diff
@@ -6318,6 +6324,17 @@ _CONSTEXPR20 void sort_heap(_RanIt _First, _RanIt _Last) { // order heap by repe
 #ifdef __cpp_lib_concepts
 namespace ranges {
     // VARIABLE ranges::sort_heap
+    // clang-format off
+    template <random_access_iterator _It, class _Pr, class _Pj>
+        requires sortable<_It, _Pr, _Pj>
+    static constexpr void _Sort_heap_common(const _It _First, _It _Last, _Pr _Pred, _Pj _Proj) {
```
Comment (CaseyCarter) [inline stl/inc/algorithm:6330]: Ditto "Nope `static`"

File: stl/inc/algorithm (hunk 4 above)
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/algorithm:5888]: Commented all three occurrences of `_Proj, _Proj`.

File: tests/std/tests/GH_001086_partial_sort_copy/test.cpp [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        x = w.x;
+        return *this;
+    }
+};
+
+int main() {
+    constexpr int src_size = 4;
+    source src[src_size];
+    constexpr int dst_size = 2;
+    target dst[2];
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_001086_partial_sort_copy/test.cpp:53]: For consistency:
```suggestion
    target dst[dst_size];
```

**Comment by StephanTLavavej** [comment]
Thanks for finding and fixing this bug! :beetle:

**Pull request merged**